File: leetcode/leetcode_cpp/design-add-and-search-words-data-structure.cpp
/*
Solution1. In place array

Time:
addWord: o(n) N is the length of the word being added
search(string word): o(m) M is the total number of nodes in the Trie.
Space: O(M), where M is the total number of nodes in the Trie
*/

class WordDictionary {
public:
    WordDictionary() {
        
    }
    
    void addWord(string word) {
        WordDictionary* node = this;
        for (auto c : word) {
            c -= 'a';
            if (!node->children[c]) node->children[c] = new WordDictionary();
            node = node->children[c];
        }
        node->isLeaf = true;        
    }
    
    bool search(string word) {
        return find(word, 0);
    }

    bool find(string word, int i) {
        if (i >= word.size()) return isLeaf;
        auto c = word[i];
        if (c != '.') {
            c -= 'a';
            if (!children[c]) return false;
            return children[c]->find(word, i+1);
        }

        for (auto child : children) {
            if (child && child->find(word, i+1))
                return true;
        }
        return false;
    }

    WordDictionary* children[26] = {};
    bool isLeaf = false;
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: leetcode/leetcode_py/design-add-and-search-words-data-structure.py
'''
Solution1. In place array

Time:
addWord: o(n) N is the length of the word being added
search(string word): o(m) M is the total number of nodes in the Trie.
Space: O(M), where M is the total number of nodes in the Trie
'''
class WordDictionary:

    def __init__(self):
        self.children = {}
        self.isLeaf = False
        
    def addWord(self, word: str) -> None:
        node = self
        for c in word:
            if not c in node.children:
                node.children[c] = WordDictionary()
            node = node.children[c]
        node.isLeaf = True

    def search(self, word: str) -> bool:                    
        return self.find(word, 0)
        
    def find(self, word, i):
        if i >= len(word):
            return self.isLeaf
        c = word[i]
        if c != '.':
            if not c in self.children:
                return False
            return self.children[c].find(word, i+1)
        else:
            for c in self.children:
                if self.children[c].find(word, i+1):
                    return True
        return False


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)